components/style/sharing/mod.rs

```diff
@@ -32,6 +34,88 @@ pub enum StyleSharingBehavior {
     Disallow,
 }
 
+/// Some data we want to avoid recomputing all the time while trying to share
+/// style.
+///
+/// This data structure is shared between the `ElementSharingChecker` and the
+/// `StyleSharingCandidate` (indeed we could try to merge those, though
+/// conceptually we may still want them to be separate).
+#[derive(Debug)]
+pub struct CachedStyleSharingData {
+    /// The class list of the element.
+    class_list: Option<SmallVec<[Atom; 5]>>,
+    /// The presentational hints this element generated.
+    pres_hints: Option<SmallVec<[ApplicableDeclarationBlock; 5]>>,
+    /// The revalidation selectors matching result.
+    revalidation_match_results: Option<BitVec>,
+}
+
+impl CachedStyleSharingData {
+    /// Create an empty `CachedStyleSharingData`.
+    pub fn new() -> Self {
+        Self {
+            class_list: None,
+            pres_hints: None,
+            revalidation_match_results: None,
+        }
+    }
+
+    /// Swap the cached data, leaving `self` empty, and returning a
+    /// `CachedStyleSharingData` with the current cache.
+    pub fn take(&mut self) -> Self {
+        Self {
+            class_list: self.class_list.take(),
+            pres_hints: self.pres_hints.take(),
+            revalidation_match_results: self.revalidation_match_results.take(),
+        }
+    }
+
+    /// Compute the class list of this element if needed, and return it.
+    fn class_list<E>(&mut self, element: E) -> &[Atom]
+        where E: TElement,
+    {
+        if self.class_list.is_none() {
+            let mut ret = SmallVec::new();
+            element.each_class(|c| ret.push(c.clone()));
+            self.class_list = Some(ret);
```

Comment:
Can we sort this list please? Otherwise we may falsely reject candidates based on different ordering.